rfcs/0015-const-generics.md

```diff
@@ -0,0 +1,299 @@
+
+```rust
+fn f<const I: usize>() {
+    __log(I);
+}
+```
+
+Different from type generics, const generics cannot appear at:
+
+1. constraints;
+1. where types are expected.
+
+```rust
+// 1 - INVALID
+fn f<const I: usize>() where I > 10 {
+    ...
+}
+
+// 2 - INVALID
+fn f<const I: usize>() -> Vec<I> {
+    Vec::<I>::new()
+}
+```
+
+## Const Value Specialization
+
+By the nature of `impl` declarations, it is possible to specialize for some specific types. For example:
+
+```rust
+impl SomeStruct<bool> {
+    fn f() {        
+ }
+}
+```
+
+In the example above, `f` only is available when the generic argument of `SomeStruct` is known to be `bool`. This will not be supported for "const generics", which means that the example below will not be supported:
+
+```rust
+impl SomeIntegerStruct<1> {
+    ...
+}
+
+impl SomeBoolStruct<false> {
+    ...
+}
+```
+
+The main reason for forbidding this is that apart from `bool`, which only needs two values, all other constants would demand complex syntax to guarantee the completeness and uniqueness of all implementations, the same way that `match` expressions do.
+
+## Monomorphization
+
+As with other generic arguments, `const generics` monormorphize functions, which means that a new "TyFunctionDecl", for example, will be created for which value that is instantiated.
+
+Prevention of code bloat will be the responsability of the optimizer.
+
+Monomorphization for const generics has one extra complexity to support arbitrary expressions it is needed to "solve" an equation. For example,
+if a variable is typed as `[u64; 1]` and a method with its `self` type as `[u64; N + 1]` is called, the monomorphization process needs to know that `N` needs to be valued as `0` and if the variable is `[u64; 2]`, `N` will be `1`.
+
+## Type Engine changes
+
+By the nature of `const generics`, it will be possible to write expressions inside types. Initially, only simple references will be
+supported, but at some point, more complex expressions will be needed, for example:
+
+```sway
+fn len<T, const N: u64>(a: [T; N]) { ... }
+
+fn bigger<const N: usize>(a: [u64; N]) -> [u64; N + 1] {
+ [0; N + 1]
+}
+```
+
+This poses the challenge of having an expression tree inside the type system. Currently `sway` already 
+has three expression trees: `Expr`, `ExpressionKind` and `TyExpressionVariant`. This demands a new one,
+given that the first two allow much more complex expressions than what `const generics` wants to support; 
+and  last one is only create after some `TypeInfo` already exist, and thus cannot be used.
+
+At the same time, the parser should be able to parse any expression and return a friendly error that such expression,
+is not supported. 
+
+So, in case of a unsupported expression the parser will parser the `const generic` expression as it does for normal `Expr`and will lower it to the type system expression enum, but in the place of the unsupported expression will return a `TypeInfo::ErrorRecovery`.
```

Comment:
```suggestion
So, in case of an unsupported expression the parser will parser the `const generic` expression as it does for normal `Expr`and will lower it to the type system expression enum, but in the place of the unsupported expression will return a `TypeInfo::ErrorRecovery`.
```